build images: deprecate SBOM attachments and use SBOM attestations (#31027)
cilium-team/cilium: Documentation/configuration/sbom.rst

Download SBOM
=============

> You can download SBOM from the supplied Cilium image using

Reviewer: ```suggestion
You can download the SBOM from the supplied Cilium image using the following command:
```

.. code-block:: shell-session

    $ cosign download attestation --predicate-type spdxjson --output-file cilium_SBOM.spdx.json <Image URI>

Verify SBOM attestation
=======================

> To verify SBOM attestation on the supplied Cilium image, run

Reviewer: ```suggestion
To verify the SBOM attestation for the supplied Cilium image, run the following command:
```

Also, it would be nice to add a sentence on what the SBOM “attestation” provides, as not all readers (or reviewers :innocent:) will be familiar with the concept. I suppose it's a signature to prove the authenticity and the integrity of the SBOM, right?

Author: > Also, it would be nice to add a sentence on what the SBOM “attestation” provides

I agree.

> To verify the SBOM attestation *for* the supplied Cilium image, run the following command:

I am following the command doc [here](https://github.com/sigstore/cosign/blob/main/doc/cosign_verify-attestation.md). I think there can be multiple attestations on an image.

Download SBOM
=============

You can download the SBOM in-toto attestation from the supplied Cilium image using the following command:

.. code-block:: shell-session

    $ cosign download attestation --predicate-type spdxjson --output-file cilium_SBOM.spdx.json <Image URI>

Verify SBOM attestation
=======================

To verify the SBOM in-toto attestation on the supplied Cilium image, run the following command:

.. code-block:: shell-session

>     $ cosign verify-attestation --certificate-github-workflow-repository cilium/cilium --certificate-oidc-issuer https://token.actions.githubusercontent.com --type spdxjson <Image URI> | jq

Reviewer: I think it would be useful to precise what the user should expect from the output (is it still the `Issuer` and `Subject` fields that the user should look at?).

Looks OK from my side otherwise, thanks! Although I'm still unfamiliar with cosign, and will rely on André to validate the changes.

Author: > I think it would be useful to precise what the user should expect from the output (is it still the Issuer and Subject fields that the user should look at?).

I agree (yes).

Download SBOM
=============

You can download the SBOM in-toto attestation from the supplied Cilium image using the following command:

.. code-block:: shell-session

    $ cosign download attestation --predicate-type spdxjson <Image URI> | jq -r .payload | base64 -d | jq .predicate > ciliumSBOM.spdx.json

Verify SBOM attestation
=======================

To verify the SBOM in-toto attestation on the supplied Cilium image, run the following command:

.. code-block:: shell-session

>     $ TAG=v1.15.0

Reviewer 2: I noticed that this doesn't work for the v1.15.0 TAGs. Can we change this so the value is automatically updated whenever we generate a new tag starting from 1.16 pre-release? cc @qmonnet

Reviewer: So we could have something like this:

```
.. parsed-literal::

    $ TAG=|CURRENT_RELEASE|
    $ cosign ...
```

But this would pick up the version from file `VERSION`, which does not correspond to an existing tag on `main`.

I'm not sure how we'd update this to pick up the latest tag. I've struggled more than I expected, recently, to figure out a way to pick up the “latest release” tag for the upgrade/downgrade workflows in CI. The issue is that there's no easy way to tell what's the relevant tag; and it also requires the Git history to be available.

Or did you have a specific solution in mind? Or would just using `|CURRENT_VERSION|` work for the docs? After all, we don't need all tags to appear in the docs?

Reviewer 2: > Or did you have a specific solution in mind? Or would just using |CURRENT_VERSION| work for the docs? After all, we don't need all tags to appear in the docs?

@qmonnet I meant the same thing we do in 
![image](https://github.com/cilium/cilium/assets/5714066/6c8c3446-d0cb-45f5-82b4-ff866a718943)

from https://docs.cilium.io/en/v1.15/gettingstarted/k8s-install-default/

Reviewer: OK so that's:

```
       .. parsed-literal::

          cilium install |CHART_VERSION|
```

With `chart_version = '--version ' + release`; so that's likely `|IMAGE_TAG|` that we'd need, given that Documentation/conf.py defines `image_tag = 'v' + release`.

    $ cosign verify-attestation --certificate-github-workflow-repository cilium/cilium \
    --certificate-oidc-issuer https://token.actions.githubusercontent.com \
    --certificate-identity "https://github.com/cilium/cilium/.github/workflows/build-images-releases.yaml@refs/tags/${TAG}" \

>     --type spdxjson <Image URI> | jq

Reviewer 2: The json output is too big, can we add `  2>&1   | head -n 13` instead of `| jq`?

For example:

```
--type spdxjson quay.io/cilium/cilium-ci:e5495c548abd25b6672431c02bb6c0460925baff  2>&1   | head -n 13

Verification for quay.io/cilium/cilium-ci:e5495c548abd25b6672431c02bb6c0460925baff --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - The code-signing certificate was verified using trusted certificate authority certificates
Certificate subject: https://github.com/cilium/cilium/.github/workflows/build-images-ci.yaml@refs/pull/31028/merge
Certificate issuer URL: https://token.actions.githubusercontent.com
GitHub Workflow Trigger: pull_request
GitHub Workflow SHA: 89175f13a4a0d6b79e06902e3e6a6ab55358001a
GitHub Workflow Name: Image CI Build
GitHub Workflow Repository: cilium/cilium
GitHub Workflow Ref: refs/pull/31028/merge
```

> It can be validated that the image was signed using Github Actions in the Cilium

Reviewer 2: I would suggest to add a "real example" in here. So that people know what they are expecting to see:
```
cosign verify-attestation --certificate-github-workflow-repository cilium/cilium     --certificate-oidc-issuer https://token.actions.githubusercontent.com     --certificate-identity https://github.com/cilium/cilium/.github/workflows/build-images-ci.yaml@refs/pull/31028/merge     --type spdxjson quay.io/cilium/cilium-ci:e5495c548abd25b6672431c02bb6c0460925baff  2>&1 | head -n 13

Verification for quay.io/cilium/cilium-ci:e5495c548abd25b6672431c02bb6c0460925baff --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - The code-signing certificate was verified using trusted certificate authority certificates
Certificate subject: https://github.com/cilium/cilium/.github/workflows/build-images-ci.yaml@refs/pull/31028/merge
Certificate issuer URL: https://token.actions.githubusercontent.com
GitHub Workflow Trigger: pull_request
GitHub Workflow SHA: 89175f13a4a0d6b79e06902e3e6a6ab55358001a
GitHub Workflow Name: Image CI Build
GitHub Workflow Repository: cilium/cilium
GitHub Workflow Ref: refs/pull/31028/merge
```

WDYT @qmonnet ?

Reviewer: Yes, I like the idea

Download SBOM
=============

The SBOM can be downloaded from the supplied Cilium image using the
``cosign download sbom`` command.

.. code-block:: shell-session

    $ cosign download sbom --output-file sbom.spdx <Image URL>

Verify SBOM Image Signature
===========================

To ensure the SBOM is tamper-proof, its signature can be verified using the
``cosign verify`` command.

.. code-block:: shell-session

    $ COSIGN_EXPERIMENTAL=1 cosign verify --certificate-github-workflow-repository cilium/cilium --certificate-oidc-issuer https://token.actions.githubusercontent.com --attachment sbom <Image URL> | jq

It can be validated that the image was signed using Github Actions in the Cilium

> repository from the ``Issuer`` and ``Subject`` fields of the output.

Reviewer: Why removing these comments? Do they no longer apply? How does one verify that the SBOM has been correctly signed?

Author: We are not attaching SBOMs as that feature is being deprecated in cosign. So, these commands to verify SBOM attachments no longer apply. `cosign attest` command automatically signs the SBOM attestations, and `cosign download attestation <image URI>` automatically verifies the SBOM [attestations](https://github.com/sigstore/cosign/blob/main/specs/ATTESTATION_SPEC.md).

Ref: 
[Cosign SBOM Specifications](https://github.com/sigstore/cosign/blob/main/specs/SBOM_SPEC.md) 
[Cosign Attestation Specifications](https://github.com/sigstore/cosign/blob/main/specs/ATTESTATION_SPEC.md)

Reviewer 2: I believe the `cosign download attestation` only downloads without verifying if they are correctly signed. Shouldn't we use `cosign verify-attestation` as well?

Author: You are right. When `cosign download sbom` command is run, it outputs:

> WARNING: Downloading SBOMs this way does not ensure its authenticity. If you want to ensure a tamper-proof SBOM, download it using 'cosign download attestation <image uri>'.

This got me thinking  `cosign download attestation <image uri>` automatically verifies an attestation on the supplied container image to ensure a tamper-proof SBOM.  We should use `cosign verify-attestation` as well.

> If you want to ensure a tamper-proof SBOM, download it using ``cosign download attestation <image uri>``.

Reviewer: ```suggestion
If you want to ensure a tamper-proof SBOM, download it using ``cosign download attestation <image URI>``.
```

Download SBOM
=============

You can download the SBOM in-toto attestation from the supplied Cilium image using the following command:

.. code-block:: shell-session

    $ cosign download attestation --predicate-type spdxjson <Image URI> | jq -r .payload | base64 -d | jq .predicate > ciliumSBOM.spdx.json

Verify SBOM attestation
=======================

To verify the SBOM in-toto attestation on the supplied Cilium image, run the following command:

.. parsed-literal::
    
    $ TAG = |IMAGE_TAG|
    $ cosign verify-attestation --certificate-github-workflow-repository cilium/cilium \
    --certificate-oidc-issuer https://token.actions.githubusercontent.com \
    --certificate-identity-regexp https://github.com/cilium/cilium/.github/workflows \
    --type spdxjson <Image URI> | 2>&1   | head -n 13

For example:

.. code-block:: shell-session

    $ cosign verify-attestation --certificate-github-workflow-repository cilium/cilium \
    --certificate-oidc-issuer https://token.actions.githubusercontent.com \
    --certificate-identity-regexp https://github.com/cilium/cilium/.github/workflows \

>     --type spdxjson quay.io/cilium/cilium-ci:e5495c548abd25b6672431c02bb6c0460925baff  2>&1 | head -n 13

Reviewer 3: ```suggestion
    --type spdxjson quay.io/cilium/cilium-ci:d2d270a42b674ca1e7c536186691d8ac8317fd64  2>&1 | head -n 13
```
